### PR TITLE
Orientation sensor extension

### PR DESCRIPTION
**Pull request by Wend1go**
This extension grants access to the gyroscope and compass sensors on mobile devices via "deviceorientation" JavaScript event handler ([trello](https://trello.com/c/kP0LxgaJ/1-support-for-accelerometer-and-gyroscope-html5-games)).
The device orientation is a combination of gyroscope, accelerometer and compass. But I could also add support for the individual sensors to the extension.

Done:
- [x] Give access to orientation sensors via JavaScript extension
- [x] Add example game: Ball parkour
- [x] Add icons for the new events
- [x] Testing exported games on Andorid / iOS

Todo:
- [ ] Add example game: Compass

I still haven't managed to get Cordova to compile an Android App for me on my local PC. It claims that I should accept some EULA which I already accepted. See my [forum post](http://compilgames.net/forum/viewtopic.php?f=18&t=10568#p67709) I'm thankful for any help I can get.
Maybe @4ian or others could compile an Android App and export a test game to check if the extension  works properly inside a Cordova App.
Running an HTML5 exported test game on different Android browsers worked pretty well for the testers on the forum.

I couldn't find the location of the icons used in other extensions, so I couldn't yet create a dedicated one for the orientation extension.

My own Android phone still runs Ice cream Sandwich and has no WebGL support, so everything is really laggy. The ballgame might need some adjustment for the movement speed.

### CODE REVIEW

**Comment by 4ian** [comment]
Wow ❤️
This looks super great. Let me look at this this evening and see what we can do for icons. 

By default Cordova is running apps on the system webview - which is Chrome for most recent Android phones (Android 5+). This means that testing on Chrome should give the same result as in a Cordova app 😄There is also [this article](https://blog.phonegap.com/migrating-from-the-cordova-device-orientation-plugin-8442b869e6cc) from the PhoneGap team explaining how to use deviceorientation, so I'm quite confident that there is good support for it on Android and iOS!

Great job! 🙌

**Comment by Wend1go** [comment]
Thanks :)
The alpha, beta and gamma values are anything but self explanatory. Would it be possible to show an image in the event editor for displaying a stylized smart phone with arrows to indicate what axis the values correspond to?

**Comment by blurymind** [comment]
This is fantastic! :D thank you @Wend1go

File: Extensions/DeviceSensors/devicesensortools.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+/**
+ * Get the value of the device orientations beta as a number
+ * @return {number} The device orientations beta value
+ */
+gdjs.evtTools.devicesensors.orientation.getOrientationBeta = function() {
+  return gdjs.evtTools.devicesensors.orientation._beta;
+};
+
+/**
+ * Get the value of the device orientations gamma as a number
```
[the remark this answers is not in the document]
Comment (Wend1go) [reply Extensions/DeviceSensors/devicesensortools.js]: I'm not really sure. It could also be "device orentation's gamma".
Are there any native English speakers around?

File: Extensions/DeviceSensors/devicesensortools.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  window.addEventListener("deviceorientation", _handleOrientation, true);
+  gdjs.evtTools.devicesensors.orientation._isActive = true;
+}
+
+function _deactivateOrientationListener() {
+  window.removeEventListener('deviceorientation', _handleOrientation, true);
+  gdjs.evtTools.devicesensors.orientation._isActive = false;
+}
+
+function _handleOrientation(event) {
```
[the remark this answers is not in the document]
Comment (Wend1go) [reply Extensions/DeviceSensors/devicesensortools.js]: Ah good to know, I always thought it was the other way around. (anything declared inside the extension's top level would be local to that extension and stuff attached to evtTools would be accessable from the outside)

[the remark this answers is not in the document]
Comment (4ian) [reply Extensions/DeviceSensors/devicesensortools.js]: @Wend1go It's the case when you're using Node.js (or bundler like Webpack, Parcel.js) but GDevelop is not using any kind of bundler and relying instead on "Vanilla" JavaScript (for simplicity sake, including a bundler is not an easy task, and also because it would slow down preview and exports).

So all code is living in the global scope, hence the need to scope everything to an object/module to avoid any name clashes.

**Comment by 4ian** [comment]
It will work for games exported with Cocos2d-JS for HTML5, but almost certainly not for mobile (android/ios) or other Cocos2d-JS native targets.

It's fine for the Cocos2D-JS export not to support this. The reasoning is that, as Cocos2D-JS export is experimental, it should support a subset of the features supported by the default GDJS game engine (i.e: HTML5 + Pixi.js).
Subset means that everything that works in Cocos2d-JS should work the same as with HTML5/Pixi.js but if something is not supported it's ok (as long as it's not breaking the game).

(Ideally we would have hints displayed in the editor to quickly verify what is supported/not support by Cocos2d-JS export - but I don't think it is worth it).

tl;dr: we don't care. 😊 But if someone want to add support or need this, any help to port the feature is welcome.

**Comment by Wend1go** [comment]
Ok the extension is ready for testing.

Its a pity that SVG images are not supported since these would have a couple of benefits:
- One size fits all / scale as much as you like
- The displayed image is also the project file (Inkscape)
(I included the SVG file anyway)

**Comment by Lizard-13** [comment]
I took the liberty of adding you :)

File: Extensions/DeviceSensors/devicesensortools.js [hunk 3]
```diff
@@ -9,18 +9,22 @@ gdjs.evtTools.devicesensors = {};
 gdjs.evtTools.devicesensors.orientation = {};
 
 gdjs.evtTools.devicesensors.orientation._isActive = false;
+gdjs.evtTools.devicesensors.orientation._absolute = 0;
+gdjs.evtTools.devicesensors.orientation._alpha = 0;
+gdjs.evtTools.devicesensors.orientation._beta =  0;
+gdjs.evtTools.devicesensors.orientation._gamma =  0;
```
Comment (4ian) [inline Extensions/DeviceSensors/devicesensortools.js]: 👍 Good catch

File: Extensions/DeviceSensors/devicesensortools.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
+/**
+ * @memberof gdjs.evtTools
+ * @class devicesensors
+ * @static
+ * @private
+ */
+
+gdjs.evtTools.devicesensors = {};
```
Comment (4ian) [inline Extensions/DeviceSensors/devicesensortools.js]: nitpicking: you can also declare an object like this:

```
gdjs.evtTools.devicesensors = {
  orientation: {
    _isActive: false,
    _absolute: 0,
    _alpha: 0;
    _beta: 0;
    _gamma: 0
  }
};
```

Slightly easier to type and read - but it's nitpicking, it's also good how you've done.

File: Extensions/DeviceSensors/JsExtension.js [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        "OrientationActive",
+        t("Sensor Active"),
+        t("The expression is true if the device orientation sensor is currently active"),
+        t("Sensors/Orientation")
+      )
+      .getCodeExtraInformation()
+      .setIncludeFile(
+        "Extensions/DeviceSensors/devicesensortools.js"
+      )
+      .setFunctionName("gdjs.evtTools.devicesensors.orientation.isActive");
```
Comment (4ian) [inline Extensions/DeviceSensors/JsExtension.js]: So, this works but it's a bit abusing the "type system" of JavaScript/GDevelop as we're returning a boolean instead of a number. This can work but also can lead to really curious error (for example, `1 + true = 2`, but converting if somehow `true` is stored in a variable, and then the variable is converted to a string, then the user will get `true` instead of 1).
So not sure if it's ideal? What you could do is have a "isActiveAsNumber" doing the conversion explicitly (something as simple as `return gdjs.evtTools.devicesensors.orientation._isActive ? 1 : 0`).

Would avoid any edge case in the future :)

**Comment by 4ian** [comment]
Tested, the example works well during the preview over wifi on my phone and also on my iPad 4 (iOS 11) :)
Note that if you want the example to be displayed in the list of examples, the name of the file `newIDE/app/resources/examples/device_orientation_ballgame/game.json` must be the same as the folder (`newIDE/app/resources/examples/device_orientation_ballgame/device_orientation_ballgame.json`). The convention is to use dash (-) to separate words (IDE replace them by space). Capital letter is added automatically to the first letter in the IDE. I will convert the example to a web-app example once it's merged. 😄 

Thanks for including the SVGs files! Support for icons in SVG would definitely be better than the 16x16 and 24x24 bitmaps. It should be easier to implement support for it now that the editor is powered by React compared to GD4 IDE. I've [created a Trello card](https://trello.com/c/kAyNjuEN/128-support-actions-conditions-icons-in-svg) to have it noted somewhere.

So I think we're almost ready to merge this! I've left one comment about expression using isActive that you may want to double check. Let me know once you're done and will merge it :)

File: Extensions/DeviceSensors/JsExtension.js (hunk 5 above)
Comment (Wend1go) [reply Extensions/DeviceSensors/JsExtension.js]: I decided to go "all number". When the events use numbers anyway there is no need for boolean inside the extension.

**Comment by Wend1go** [comment]
Done, ready to merge :)
I'll add the compass example later.

**Comment by 4ian** [comment]
Looks good to me!

Great job on this! This was something that Gdvdelop lacked a lot 😄

**Comment by Wend1go** [comment]
Thanks :)
Just added the wiki article:
[http://wiki.compilgames.net/doku.php/gdevelop5/all-features/device-sensors#conditions](http://wiki.compilgames.net/doku.php/gdevelop5/all-features/device-sensors#conditions)

**Comment by Wend1go** [comment]
While testing the auto completion, I noticed that there is an inputManager object (runtimeScene.getGame().getInputManager()) which contains keyboard, mouse and touch input functions. If you agree I would move the device sensor into the input manager since it makes more sense than having it in a separate namespace. And I'd rather do it asap before people start accessing the orientation sensor via javascript events.
How would I access the input manager object from within the DeviceOrientation extension?

**Comment by 4ian** [comment]
I'm not sure if it's preferable to do this, as we lose modularity and instead going to have InputManager dynamically modified by an extension - that sounds dangerous to me.
Wouldn't it be more simple to just have the extension exposed in its own "namespace" as now?

`gdjs.evtTools.devicesensors.orientation` is indeed a bit long and as it's inside `gdjs.evtTools`, it's usually meant to be used by events only.
But we can then move it to `gdjs.deviceSensors.orientation` (note the capital S, I should have seen this before - convention is to use camelCase in JavaScript/in the game engine) so that it's nicer to use from JS. Attaching it to InputManager does not provide more value though as far as I can see. Do you have something else in mind?

**Comment by Wend1go** [comment]
Well I just thought it would be easier to find if it was located in the input section.
The structure of gdjs is currently pretty flat. Objects, Behaviours, Math functions, Renderers and other stuff is all located in the same namespace while other functionality can only be found under gdjs.evtTools. It isn't very intuitive currently.

**Comment by 4ian** [comment]
> other functionality like the orientation can only be found under gdjs.evtTools

Yep, that is what I was saying: things that are in `gdjs.evtTools` should only be normally accessed by events, and are not meant to be public facing (i.e: used by anyone in JavaScript other than the extension itself).

The idea is too keep a flat structure: everything that is in gdjs should normally be "pretty usable" by everyone (gdjs.RuntimeXXXObject, gdjs.RuntimeXXXBehavior, gdjs.RuntimeScene), so:
* If you have something that you want to be accessible from JavaScript, add it directly under your own namespace in gdjs: `gdjs.deviceSensors`.
* If anything should be available for events AND JavaScript code, you can leave it in `gdjs.deviceSensors`
* If anything should be available only for events, you can put it in `gdjs.evtTools.deviceSensors` (but it's a convention)
* If for some reason there already are things that are in gdjs but should not be public, let's add `@private` annotation.

Does that make sense?

**Comment by 4ian** [comment]
Also in the case of your extension, it's ok to be in `gdjs` directly because orientation are "global" values that does not depends on the scene or anything.
`InputManager` does not depends on the scene so that's why it is in `gdjs.RuntimeGame` (it could even have been a singleton object in gdjs, without needing to do runtimeScene.getGame().getInputManager()).

Other extensions that deal with objects or behaviors need to keep everything inside objects or behaviors. In your case, it's ok to have a unique JS object holding your extension.

**Comment by Wend1go** [comment]
Ok I will put the device orientation in `gdjs.deviceSensors` then, since the functions are save to use in JavaScript events. And I'll add `@private` to the private functions.

**Comment by 4ian** [comment]
@Wend1go Cool that sounds good and safely reusable in JavaScript 👍

**Comment by Wend1go** [comment]
I can't hide the private functions from Monacos code completion. Am I doing something wrong or is it even possible to hide it?
Currently I can only see the `@private` notice when I already wrote the complete function call and hover the mouse over its function name.
```
 /**
  * Activate the orientation sensor's listener.
  * @private
 */
gdjs.deviceSensors.orientation._handleOrientation = function(event) {
  gdjs.deviceSensors.orientation._absolute = event.absolute ? Math.round(event.absolute) : 0;
  gdjs.deviceSensors.orientation._alpha = event.alpha ? Math.round(event.alpha) : 0;
  gdjs.deviceSensors.orientation._beta = event.beta ? Math.round(event.beta) : 0;
  gdjs.deviceSensors.orientation._gamma = event.gamma ? Math.round(event.gamma) : 0;
}
```

**Comment by 4ian** [comment]
I think even `@private` won't hide in auto completion indeed. But having the underscore in front of the name is good hint (well known in JS) that it should not be used from outside the extension, so I think it's good enough.

**Comment by Wend1go** [comment]
I added the motion sensor to the extension but there is a little problem.
While the acceleration along the x, y and z axis works good, the rotation around the axes seems to be badly supported on mobile devices.

Possibilities:
1) Remove the rotation code to not confuse people with sensor functionality which in most cases doesn't deliver any values
2) Leave it in the extension but comment it out in case things change in the future. (It feels bad to remove existing code)
3) Leave it active in the extension (and maybe write a hint, about bad support on mobile phones). Who knows maybe someone uses GD on a Raspberry Pi with such a sensor attached for education. 

What do you think @4ian ?

**Comment by 4ian** [comment]
Let's keep it as it could allow creative usage when it's properly supported, but let's add some explicit warning in the description (and also in the help page) about the bad compatibility of most devices with this (so it should either be avoided or tested a lot).
